### PR TITLE
Add clock stop & jack detection support for SDW

### PR DESCRIPTION
**Pull request by RanderWang**
on intel platform, a glue HW helps clock stop and jack detection. The clock stop steps are:
(1) Master & slave are  set to clock stop mode. 
(2) Glue HW takes over soundwire bus. WAKEEN feature is set for wake-up event like jack event
(3)  audio pci device will go to D3 state
(4)  When slave detects jack event, it will send a signal to notify Glue HW to trigger a PME event to ACPI system to resume audio pci driver to D0 state
(5) SOF driver check WAKEEN status and resume master & slave to check the jack status.

### CODE REVIEW

File: drivers/soundwire/bus.c [hunk 1]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+			val |= SDW_SCP_SYSTEMCTRL_WAKE_UP_EN;
+	} else {
+		val = sdw_read_no_pm(slave, SDW_SCP_SYSTEMCTRL);
+
+		val &= ~(SDW_SCP_SYSTEMCTRL_CLK_STP_PREP);
+	}
+
+	ret = sdw_write_no_pm(slave, SDW_SCP_SYSTEMCTRL, val);
+
+	if (ret != 0 && ret != -ENODATA)
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: the second part is questionable.
I believe ret = -ENODATA means the device did not respond. I am not sure we want to go ahead even in this case, or it would be to be explicit with a comment.

File: drivers/soundwire/bus.c [hunk 2]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+		val = sdw_bread_no_pm(bus, dev_num,
+				      SDW_SCP_STAT & SDW_SCP_STAT_CLK_STP_NF);
+		if (val <= 0)
+			break;
+
+		usleep_range(1000, 1500);
+		retry--;
+	} while (retry);
+
+	if (retry && (val == 0 || val == -ENODATA)) {
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: same here, not sure why -ENODATA is an acceptable reply on a broadcast read?

File: drivers/soundwire/bus.c [hunk 3]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+	 * prepared for clock stop.
+	 */
+	list_for_each_entry(slave, &bus->slaves, node) {
+		if (!slave->dev_num)
+			continue;
+
+		/* Identify if Slave(s) are available on Bus */
+		is_slave = true;
+
+		if (slave->status == SDW_SLAVE_ATTACHED) {
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: if the slave status changes to ALERT, what happens?
This test is not good enough.
I think it meant ATTACHED || ALERT

File: drivers/soundwire/bus.c [hunk 4]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+		if (ret < 0)
+			return ret;
+	}
+
+	/* Inform slaves that prep is done */
+	list_for_each_entry(slave, &bus->slaves, node) {
+		if (!slave->dev_num)
+			continue;
+
+		if (slave->status != SDW_SLAVE_ATTACHED)
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: ATTACHED || ALERT

File: drivers/soundwire/bus.c [hunk 5]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+			continue;
+
+		slave_mode = slave->curr_clk_stop_mode;
+
+		if (!slave->prop.simple_clk_stop_capable) {
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: I am not sure why this part is conditional.
It could very well be that the Slave driver only implements a POST_PREPARE callback.

File: drivers/soundwire/bus.c [hunk 6]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+{
+	int ret;
+
+	/*
+	 * broadcast clock stop now, attached Slaves will ACK this,
+	 * unattached will ignore
+	 */
+	ret = sdw_bwrite_no_pm(bus, SDW_BROADCAST_DEV_NUM,
+			       SDW_SCP_CTRL, SDW_SCP_CTRL_CLK_STP_NOW);
+	if (ret != 0 && ret != -ENODATA) {
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: this again makes no sense. If the -ENODATA is returned, it means the command was ignored.
With the Cadence IP, I think the clock would not stop here. This needs to be checked.

File: drivers/soundwire/bus.c [hunk 7]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+	 * de-prepared after clock resume.
+	 */
+	list_for_each_entry(slave, &bus->slaves, node) {
+		if (!slave->dev_num)
+			continue;
+
+		/* Identify if Slave(s) are available on Bus */
+		is_slave = true;
+
+		if (slave->status != SDW_SLAVE_ATTACHED)
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: could be ALERT as well

File: drivers/soundwire/bus.c [hunk 8]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+			continue;
+
+		mode = slave->curr_clk_stop_mode;
+
+		if (mode == SDW_CLK_STOP_MODE1)
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: actually if we resume in clock_stop_mode1 then we need to clear the status and indicate it so that the Slave resume will wait for the enumeration/initialization to complete.

File: drivers/soundwire/bus.c [hunk 9]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+	}
+
+	if (is_slave && !simple_clk_stop)
+		sdw_bus_wait_for_clk_prep_deprep(bus, SDW_BROADCAST_DEV_NUM);
+
+	list_for_each_entry(slave, &bus->slaves, node) {
+		if (!slave->dev_num)
+			continue;
+
+		if (slave->status != SDW_SLAVE_ATTACHED)
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: or ALERT

File: drivers/soundwire/bus.c [hunk 10]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+			continue;
+
+		mode = slave->curr_clk_stop_mode;
+		if (mode == SDW_CLK_STOP_MODE1)
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: this is also problematic.
By the time we 'deprep' the devices, the ones that were in clock-stop-mode1 may have rejoined the bus and be ATTACHED. I am not sure why we continue here.

File: include/linux/soundwire/sdw.h [hunk 11]
```diff
@@ -910,6 +931,9 @@ struct sdw_stream_runtime {
 
 struct sdw_stream_runtime *sdw_alloc_stream(const char *stream_name);
 void sdw_release_stream(struct sdw_stream_runtime *stream);
+
+int sdw_compute_params(struct sdw_bus *bus);
+
```
Comment (plbossart) [inline include/linux/soundwire/sdw.h]: bad merge/squash? Why is this here?

File: drivers/soundwire/cadence_master.c [hunk 12]
```diff
@@ -1178,6 +1186,133 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	 * into a state in which it ignores wake-up trials in clock
+	 * stop state
+	 */
+	if (block_wake)
+		cdns_updatel(cdns, CDNS_MCP_CONTROL,
+			     CDNS_MCP_CONTROL_BLOCK_WAKEUP,
+			     CDNS_MCP_CONTROL_BLOCK_WAKEUP);
+
+	/* Disable cmd accept mode */
+	cdns_updatel(cdns, CDNS_MCP_CONTROL, CDNS_MCP_CONTROL_CMD_ACCEPT, 0);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c]: that one is odd, since the bus clock stop is a broadcast write command. No sure if this belongs here?

File: drivers/soundwire/cadence_master.c [hunk 13]
```diff
@@ -1178,6 +1186,133 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	/* Enter clock stop */
+	ret = sdw_bus_clk_stop(&cdns->bus);
+	if (ret < 0) {
+		dev_err(cdns->dev, "bus clock stop failed %d", ret);
+		return ret;
+	}
+
+	ret = cdns_set_wait(cdns, CDNS_MCP_STAT,
+			    CDNS_MCP_STAT_CLK_STOP,
+			    CDNS_MCP_STAT_CLK_STOP);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1258]: Ah, maybe this is needed after all.
The command is send and if it returns successfullly, its effect only happens at the start of the next frame, so there is potentially a risk to if we didn't wait the bus does something silly with a clock still running.

File: drivers/soundwire/cadence_master.c [hunk 14]
```diff
@@ -1178,6 +1186,133 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	/* flush command FIFOs */
+	cdns_updatel(cdns, CDNS_MCP_CONTROL, CDNS_MCP_CONTROL_CMD_RST,
+		     CDNS_MCP_CONTROL_CMD_RST);
+
+	/* commit changes */
+	cdns_update_config(cdns);
+
+	ret = cdns_set_wait(cdns, CDNS_MCP_STAT, CDNS_MCP_STAT_CLK_STOP, 0);
+	if (ret < 0) {
+		dev_err(cdns->dev, "master clock stop failed %d\n", ret);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c]: clock stop exit failed?

File: drivers/soundwire/cadence_master.c [hunk 15]
```diff
@@ -1178,6 +1186,133 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	cdns_updatel(cdns, CDNS_MCP_CONFIG,
+		     CDNS_MCP_CONFIG_OP,
+		     CDNS_MCP_CONFIG_OP_NORMAL);
+
+	/* flush command FIFOs */
+	cdns_updatel(cdns, CDNS_MCP_CONTROL, CDNS_MCP_CONTROL_CMD_RST,
+		     CDNS_MCP_CONTROL_CMD_RST);
+
+	/* commit changes */
+	cdns_update_config(cdns);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1305]: the intent was that some of those initializations are handed in cdns_init()

Not sure why you moved everything in  cdns_resume, it makes it harder to see what's done in clock mode and what isn't

File: drivers/soundwire/intel.c [hunk 16]
```diff
@@ -1590,7 +1604,7 @@ static int intel_resume_runtime(struct device *dev)
 		return ret;
 	}
 
-	ret = sdw_cdns_exit_reset(cdns);
+	ret = sdw_cdns_resume(cdns, false);
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: well it depends. We have a quirk to disable clock-stop-mode even in pm_runtime cases.

File: drivers/soundwire/intel.c [hunk 17]
```diff
@@ -1549,70 +1527,66 @@ static int intel_resume(struct device *dev)
...
+	int link_flags;
+	bool clock_stop = true;
 
-	dev_err(dev, "%s done\n", __func__);
+	link_flags = md_flags >> (sdw->cdns.bus.link_id * 8);
+	if (link_flags & SDW_INTEL_MASTER_DISABLE_CLOCK_STOP)
+		clock_stop = false;
 
-	return ret;
+	return _resume(dev, clock_stop);
```
Comment (plbossart) [inline drivers/soundwire/intel.c:1639]: this is really hard to read, we should keep resume and runtime_resume separate until they do what we want, then we combine them.

File: drivers/soundwire/intel.c [hunk 18]
```diff
@@ -1589,18 +1589,21 @@ static int intel_resume_runtime(struct device *dev)
...
+#else
+#define INTEL_PM	NULL
+#endif
+
 struct sdw_md_driver intel_sdw_driver = {
 	.driver = {
 		.name = "intel-sdw",
 		.owner = THIS_MODULE,
-		.pm = &intel_pm,
+		.pm = INTEL_PM,
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: no, this should be done with the routines compiled out and a fallback provided.

File: drivers/soundwire/intel.c [hunk 19]
```diff
@@ -401,6 +401,41 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
...
+		if (!(wake_sts & BIT(sdw->instance)))
+			continue;
+
+		/*
+		 * wake up master and slave so that slave can notify master
+		 * the jack event and let codec driver check jack status
+		 */
+		list_for_each_entry(slave, &bus->slaves, node) {
+			if (slave->prop.wake_capable) {
+				if (slave->status != SDW_SLAVE_ATTACHED)
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: could be in alert already?

File: drivers/soundwire/intel.c [hunk 20]
```diff
@@ -401,6 +401,41 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
...
+					continue;
+
+				pm_request_resume(&slave->dev);
+			}
+		}
+	}
+}
+EXPORT_SYMBOL(sdw_intel_jack_detect);
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: where is this used?

Comment (RanderWang) [reply drivers/soundwire/intel.c]: It is used in SOF. But Now I find there is no sdw change in sof driver, so I can't check it in.

File: drivers/soundwire/intel.c (hunk 19 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: our ACPI table reports 700, 701 & 711 on CML, so these slaves will be added but only 711 is attached.

Comment (RanderWang) [reply drivers/soundwire/intel.c]: compared to normal clock stop mode, master is powered off in our clock stop implementation, so master can't response to slave. Instead, glue hw will get the wake-up event and report a PME event to PCI to wake up pci driver (sof-audio).   sof-audio driver will wake up master and slave to check event

File: drivers/soundwire/intel.c (hunk 18 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: thanks, I will refine it.

File: drivers/soundwire/cadence_master.c (hunk 15 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1305]: yes, it is duplicated for intel platform. I am not sure cdns_init() will be invoked for clock stop mode for non-intel platform, so I move some setting to cdns_resume.

File: drivers/soundwire/cadence_master.c (hunk 12 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c]: sorry, the comments should be not exactly correct.  According to spec, we need to set CMD_ACCEPT to true to cause frame shape change or to enter clock stop if there is no slave attached.  And if there are some slaves attached, it should be set false.

BTW since default setting in cdns_init is true, we need to set it to false if there is a slave attached, and set it to true if there is no slave attached.  This is another issue.

File: include/linux/soundwire/sdw.h (hunk 11 above)
Comment (RanderWang) [reply include/linux/soundwire/sdw.h]: sorry, it is a merging issue.

File: drivers/soundwire/bus.c (hunk 1 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: According to caller,  if ret == -ENODATA,  clock-stop will be failed. So we can just remove this check and print error message

File: drivers/soundwire/bus.c (hunk 3 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: yes,  I get it.

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: Thanks, I will refine it.

File: drivers/soundwire/bus.c (hunk 8 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: I will refine it.

**Comment by RanderWang** [comment]
@plbossart   Update my PR.
I checked spec, hsd and window driver again and got a more clear picture about clock stop on intel platform. Compared to normal clock stop,  master is powered off and glue hw takes over bus. So in resume function,  we need to initialized master as usual.

And We also need to clear WAKEEN interrupt ASAP, or it will affect IPC interrupt. Now there is no ipc issue when WAKEEN irq is cleared in WAKEEN process function

File: drivers/soundwire/bus.c (hunk 5 above)
Comment (plbossart) [reply drivers/soundwire/bus.c]: still need to remove this condition.

File: drivers/soundwire/bus.c [hunk 21]
```diff
@@ -663,6 +710,299 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+{
+	int ret;
+
+	/*
+	 * broadcast clock stop now, attached Slaves will ACK this,
+	 * unattached will ignore
+	 */
+	ret = sdw_bwrite_no_pm(bus, SDW_BROADCAST_DEV_NUM,
+			       SDW_SCP_CTRL, SDW_SCP_CTRL_CLK_STP_NOW);
+	if (ret != 0) {
```
Comment (plbossart) [inline drivers/soundwire/bus.c]: if (ret) ?

File: drivers/soundwire/bus.c [hunk 22]
```diff
@@ -663,6 +710,299 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+		if (slave->status != SDW_SLAVE_ATTACHED &&
+		    slave->status != SDW_SLAVE_ALERT)
+			continue;
+
+		mode = slave->curr_clk_stop_mode;
+
+		if (mode == SDW_CLK_STOP_MODE1) {
+			request = SDW_UNATTACH_REQUEST_MASTER_RESET;
+			sdw_clear_slave_status(bus, request);
+			continue;
```
Comment (plbossart) [inline drivers/soundwire/bus.c:964]: sorry, I must have confused you here.
sdw_clear_slave_status() will set the status for ALL devices.
here we only want to make sure the status is reset for devices that are exiting clock_stop_mode1

I also wonder if this isn't too late already, we should mark this status *before* restarting the clock.

File: include/linux/soundwire/sdw.h [hunk 23]
```diff
@@ -910,6 +931,7 @@ struct sdw_stream_runtime {
 
 struct sdw_stream_runtime *sdw_alloc_stream(const char *stream_name);
 void sdw_release_stream(struct sdw_stream_runtime *stream);
+
```
Comment (plbossart) [inline include/linux/soundwire/sdw.h:934]: spurious newline

File: drivers/soundwire/cadence_master.c [hunk 24]
```diff
@@ -1178,6 +1186,112 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+
+	/* Check suspend status */
+	status = cdns_readl(cdns, CDNS_MCP_STAT);
+	if (status & CDNS_MCP_STAT_CLK_STOP) {
+		dev_dbg(cdns->dev, "Clock is already stopped\n");
+		return 1;
+	}
+
+	/*
+	 * In some system it is required to be able to put master
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c]: For specific platforms, it is required...

File: drivers/soundwire/cadence_master.c [hunk 25]
```diff
@@ -1178,6 +1186,112 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	 */
+	if (block_wake)
+		cdns_updatel(cdns, CDNS_MCP_CONTROL,
+			     CDNS_MCP_CONTROL_BLOCK_WAKEUP,
+			     CDNS_MCP_CONTROL_BLOCK_WAKEUP);
+
+	/*
+	 * when no slaves are attached to the bus, cmd accept
+	 * mode should be set to 1 to enter clock stop mode
+	 *
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c]: remove line

File: drivers/soundwire/cadence_master.c [hunk 26]
```diff
@@ -1178,6 +1186,112 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+	 */
+	if (list_empty(&cdns->bus.slaves))
+		cdns_updatel(cdns, CDNS_MCP_CONTROL,
+			     CDNS_MCP_CONTROL_CMD_ACCEPT, 1);
+	else
+		cdns_updatel(cdns, CDNS_MCP_CONTROL,
+			     CDNS_MCP_CONTROL_CMD_ACCEPT, 0);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1237]: I don't think the test is correct.
This CMD_ACCEPT should be used when there are no devices attached on the link. If this is not set, if there is a broadcast write then the command ignored will be treated as a failure.

This should a scan that checks that each Slave is ATTACHED or ALERT, then you can change this bit to force a command ignored to be treated as 'no objection'

File: drivers/soundwire/cadence_master.c [hunk 27]
```diff
@@ -1178,6 +1186,112 @@ static const struct sdw_master_port_ops cdns_port_ops = {
...
+		dev_err(cdns->dev, "Couldn't exit from clock stop\n");
+		return ret;
+	}
+
+	/* Enable block wakeup */
+	if (!block_wake)
+		cdns_updatel(cdns, CDNS_MCP_CONTROL,
+			     CDNS_MCP_CONTROL_BLOCK_WAKEUP, 0);
+
+	cdns_updatel(cdns, CDNS_MCP_CONTROL, CDNS_MCP_CONTROL_CMD_ACCEPT, 0);
```
Comment (plbossart) [inline drivers/soundwire/cadence_master.c:1297]: add comment on what this does. I believe you now treat all transactions on the bus as successful only if the answer is ACK.

File: drivers/soundwire/cadence_master.c (hunk 15 above)
Comment (plbossart) [reply drivers/soundwire/cadence_master.c:1305]: the idea was to call cdns_init() from cdns_resume().

File: drivers/soundwire/intel.c [hunk 28]
```diff
@@ -1590,12 +1600,23 @@ static int intel_resume_runtime(struct device *dev)
 		return ret;
 	}
 
+	/*
+	 * master is powered off on intel platforms for clock stop state,
+	 * so it should be reset for master initialization.
+	 */
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: No. exit-reset performs a HardReset which is not required when resuming from clock stop mode.
If you do a hard-reset, you force all slaves to re-enumerate and be re-initialized, so you lose all benefits of the clock stop mode.
You will still be able to resume, but it's much longer than needed, and you've lost all context.
You need to make this conditional on clock-stop being used

File: drivers/soundwire/intel.c (hunk 16 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: You really want to check if clock-stop is allowed or not as I tried to explain above.

File: drivers/soundwire/intel.c [hunk 29]
```diff
@@ -1412,6 +1412,108 @@ static int intel_master_remove(struct sdw_master_device *md)
...
+		dev_err(dev, "%s failed: %d", __func__, ret);
+		return ret;
+	}
+
+	/*
+	 * make sure all Slaves are tagged as UNATTACHED and provide
+	 * reason for reinitialization
+	 */
+	sdw_clear_slave_status(&sdw->cdns.bus,
+			       SDW_UNATTACH_REQUEST_MASTER_RESET);
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: no, this should only be done when you don't use clock-stop mode.

File: drivers/soundwire/intel.c [hunk 30]
```diff
@@ -1412,6 +1412,108 @@ static int intel_master_remove(struct sdw_master_device *md)
...
+	if (ret < 0) {
+		dev_err(dev, "cannot enable interrupts during resume\n");
+		return ret;
+	}
+
+	/*
+	 * master is powered off on intel platforms, so it
+	 * should be reset for master initialization.
+	 */
+	ret = sdw_cdns_exit_reset(cdns);
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: no. this is not needed when exiting clock stop mode.

File: drivers/soundwire/intel.c [hunk 31]
```diff
@@ -1591,12 +1591,17 @@ static int intel_resume_runtime(struct device *dev)
...
 	SET_SYSTEM_SLEEP_PM_OPS(intel_suspend, intel_resume)
 	SET_RUNTIME_PM_OPS(intel_suspend_runtime, intel_resume_runtime, NULL)
 };
+#else
+static const struct dev_pm_ops intel_pm = {
+	SET_SYSTEM_SLEEP_PM_OPS(NULL, NULL)
+	SET_RUNTIME_PM_OPS(NULL, NULL, NULL)
+};
+
+#endif
```
Comment (plbossart) [inline drivers/soundwire/intel.c]: no. you can use fallback functions for intel_suspend/resume, etc which return 0 when PM runtime is not needed.

File: drivers/soundwire/intel.c (hunk 19 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: it doesn't matter what the master does.
The point is that the slave can report in ALERT from the first PING frame, this is a different state machine. The master cannot control what the slave does.

File: drivers/soundwire/intel.c (hunk 30 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: @plbossart   If this function is not invoked,  I got IO transfer timeout.  Master is powered off and it can't maintain its context. This is different with clock stop in SDW spec.  please check my summary

File: drivers/soundwire/intel.c (hunk 29 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: please check the following explanation for sdw_cdns_exit_reset. Without this function, I got IO timeout

File: drivers/soundwire/bus.c (hunk 21 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: yes, I will refine it

File: drivers/soundwire/bus.c (hunk 22 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c:964]: please check my explanation about clock stop state on intel platform,  we need to reset the bus since master is initialized from power-off state . Since sdw_clear_slave_status is done in resume function, this one is duplicated and too late.  I will remove it.

File: drivers/soundwire/cadence_master.c (hunk 24 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c]: thanks, I will refine it

File: drivers/soundwire/cadence_master.c (hunk 25 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c]: thanks, I will remove it

File: drivers/soundwire/cadence_master.c (hunk 26 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1237]: yes, good catch! I will refine it.

File: drivers/soundwire/cadence_master.c (hunk 27 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:1297]: thanks, I will update it

File: drivers/soundwire/intel.c (hunk 16 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: thanks, I will move this check from next commit to here.

File: drivers/soundwire/intel.c (hunk 31 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: ok, it is better to do it in another PR.

File: drivers/soundwire/intel.c (hunk 19 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: yes, I get it.

**Comment by RanderWang** [comment]
update my PR, thanks!

File: drivers/soundwire/bus.c [hunk 32]
```diff
@@ -663,6 +710,296 @@ static void sdw_modify_slave_status(struct sdw_slave *slave,
...
+}
+
+static int sdw_bus_wait_for_clk_prep_deprep(struct sdw_bus *bus, u16 dev_num)
+{
+	int retry = bus->clk_stop_timeout;
+	int val;
+
+	do {
+		val = sdw_bread_no_pm(bus, dev_num,
+				      SDW_SCP_STAT & SDW_SCP_STAT_CLK_STP_NF);
```
Comment (bardliao) [inline drivers/soundwire/bus.c]: @RanderWang This doesn't looks right. Shouldn't it be 
```val = sdw_bread_no_pm(bus, dev_num, SDW_SCP_STAT) & SDW_SCP_STAT_CLK_STP_NF;```?

File: drivers/soundwire/cadence_master.c [hunk 33]
```diff
@@ -222,6 +222,24 @@ static int cdns_clear_bit(struct sdw_cdns *cdns, int offset, u32 value)
...
+{
+	int timeout = 10;
+	u32 reg_read;
+
+	/* Wait for bit to be set */
+	do {
+		reg_read = readl(cdns->registers + offset);
+		if ((reg_read & mask) == value)
+			return 0;
+
```
Comment (bardliao) [inline drivers/soundwire/cadence_master.c:242]: Will it be better to return -ETIMEDOUT on line 247 ?

File: include/linux/soundwire/sdw.h (hunk 23 above)
Comment (RanderWang) [reply include/linux/soundwire/sdw.h:934]: removed

File: drivers/soundwire/cadence_master.c (hunk 33 above)
Comment (RanderWang) [reply drivers/soundwire/cadence_master.c:242]: yes, I prefer -ETIMEDOUT

File: drivers/soundwire/bus.c (hunk 32 above)
Comment (RanderWang) [reply drivers/soundwire/bus.c]: yes, good catch

File: drivers/soundwire/intel.c (hunk 29 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: NO. you cannot use this 
````c
	/* use hardware generated reset */
	cdns_updatel(cdns, CDNS_MCP_CONTROL,
		     CDNS_MCP_CONTROL_HW_RST,
		     CDNS_MCP_CONTROL_HW_RST);
````

it has to become dependent on the clock stop mode

File: drivers/soundwire/intel.c (hunk 30 above)
Comment (plbossart) [reply drivers/soundwire/intel.c]: same comment

File: drivers/soundwire/intel.c [hunk 34]
```diff
@@ -1412,6 +1412,96 @@ static int intel_master_remove(struct sdw_master_device *md)
 
 #ifdef CONFIG_PM
 
+static int _suspend(struct device *dev, bool clock_stop)
+{
+	struct sdw_cdns *cdns = dev_get_drvdata(dev);
+	struct sdw_intel *sdw = cdns_to_intel(cdns);
+	int ret;
+
+	dev_err(dev, "%s start\n", __func__);
```
Comment (ranj063) [inline drivers/soundwire/intel.c]: dev_err?

File: drivers/soundwire/intel.c [hunk 35]
```diff
@@ -1412,6 +1412,96 @@ static int intel_master_remove(struct sdw_master_device *md)
...
+	ret = intel_link_power_down(sdw);
+	if (ret) {
+		dev_err(dev, "Link power down failed: %d", ret);
+		return ret;
+	}
+
+	/* set WAKEEN interrupt for wake-up events */
+	intel_shim_wake(sdw, clock_stop);
+
+	dev_err(dev, "%s done\n", __func__);
```
Comment (ranj063) [inline drivers/soundwire/intel.c]: same comment for all dev_err's in this patch

File: drivers/soundwire/intel.c [hunk 36]
```diff
@@ -401,6 +401,45 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
 	}
 }
 
+void sdw_intel_process_wakeen_event(struct sdw_intel_ctx *ctx)
```
Comment (ranj063) [inline drivers/soundwire/intel.c:404]: typo in commit message "tigger a"->"trigger an"

File: drivers/soundwire/intel.c [hunk 37]
```diff
@@ -401,6 +401,45 @@ static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
...
+		struct sdw_intel *sdw;
+		u16 wake_sts;
+
+		md = link->md;
+		bus = &link->cdns->bus;
+		sdw = cdns_to_intel(link->cdns);
+		wake_sts = intel_readw(sdw->link_res->shim, SDW_SHIM_WAKESTS);
+
+		/* disable WAKEEN interrupt ASAP to prevent interrupt flood */
+		intel_shim_wake(sdw, false);
```
Comment (ranj063) [inline drivers/soundwire/intel.c]: @RanderWang just wondering: do the above 2  statements to read the wake_sts and disable wakeen need to be inside the list iterator or can they be outside?

Comment (RanderWang) [reply drivers/soundwire/intel.c]: @ranj063  intel_shim_wake(sdw, false) is for each link,  so it can't be outside.
 Yes, It is better to move wake_sts outsize. I found it was easy to get shim with sdw->link_res->shim in inside.

 I will move it outside.

File: drivers/soundwire/intel.c (hunk 36 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c:404]: thanks! I will fix it.

File: drivers/soundwire/intel.c (hunk 34 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: thanks, dev_dbg ?  I will refine it.

File: drivers/soundwire/intel.c (hunk 35 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: thanks! I will refine it.

File: drivers/soundwire/intel.c (hunk 29 above)
Comment (RanderWang) [reply drivers/soundwire/intel.c]: thanks, I will refine it.

Comment (RanderWang) [reply drivers/soundwire/intel.c]: I will try the merging irq patch, maybe it will improve my issue.

**Comment by RanderWang** [comment]
@plbossart  @ranj063   I update my PR according to your comments.

Now I remove the HW RST and don't clear slave status if it is in clock stop 0
I got Io timeout. @plbossart can you help to review this log ? thanks.


[sdw_io_timeout.log](https://github.com/thesofproject/linux/files/3874045/sdw_io_timeout.log)

**Comment by plbossart** [comment]
closing, all the clock stop and jack detection work happens on PR #1586
